File: data_structures_and_algorithms/Code_401/class-09/solutions/interview-01/javascript/reverseLinkedList/reverseLinkedList.js
'use strict';

function reverseLinkedList(ll) {
  let current = ll.head;
  let previous = null;

  while(current) {
    let next = current.next;
    current.next = previous;
    previous = current;
    current = next;
  }

  ll.head = previous;

  return ll;
}

module.exports = reverseLinkedList;

File: data_structures_and_algorithms/Code_401/class-09/solutions/interview-01/python/reverse_linked_list/reverse_linked_list.py
class LinkedList:
    def __init__(self, sequence=[]):
        self.head = None
        for item in sequence:
            self.insert(item)

    def insert(self, value):

        node = Node(value)

        if not self.head:
            self.head = Node(value)
            return

        current = self.head

        while current.next:
            current = current.next

        current.next = node

    def __str__(self):

        txt = ""

        current = self.head

        while current:
            txt += f"[{current.value}] -> "
            current = current.next
            if not current:
                txt += "X"

        return txt


class Node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next


def reverse_linked_list(ll):
    prev = None
    current = ll.head
    while current:
        next_node = current.next
        current.next = prev
        prev = current
        current = next_node

    ll.head = prev

    return ll

